### DISCUSSION TITLE
ESP32 NodeMCU with ILI9341 SPI

### DISCUSSION THREAD

**Discussion by sl4m01** [HASwitchPlate/openHASP#199] [Q&A] [answered]
Dear all,
Is it possible to use the following combination for OpenHASP:
- ESP32 NodeMCU Module WLAN WiFi Dev Kit C Development Board with CP2102 (SoC ESP32-WROOM 32)
- 2 4" ILI9341 SPI touch display 

Is it possible to use a pre compiled file (for example Wemos D1 mini ESP32)?

Or do I need to compile it by myself? If yes can someone give me a hint which file I do have to adapt?

Thank you :)

**Comment by fvanroie** [comment, marked as answer]
Sure! The `d1-mini-esp32_ili9341_full_4MB_v0.6.1.bin` binary should work on _any_ ESP32-WROOM _(or even ESP32-WROVER if gpio 16 and 17 are unused)_. The d1-mini-esp32 is just our ESP32 of choice and it is also compatible with other development boards, like the ESP32 NodeMCU.

The pins used in `d1-mini-esp32_ili9341_full_4MB_v0.6.1.bin` are choosen based on the d1 mini ESP32 with Lolin 2.4" TFT shield, but you can essentially use _any_ ILI9341 or ESP32 hardware instead. The pinout for that binary to connect the ESP32 to your screen is described here: https://haswitchplate.github.io/openHASP-docs/0.6.1/displays/MSPxxxx/#pin-configuration

So you don't *need* to compile the firmware yourself if you use that pinout. The difference in binaries vary from device-to-device if they need other drivers, specific pins or hardware/software configurations.

In your case, just connect the jumper wires according to the documentation and use `d1-mini-esp32_ili9341_full_4MB_v0.6.1.bin`.
> Tip: Leave `TFT_MISO` unconnected if you only get a white screen on an MSP display, that pin it's not used in openHASP.

**Reply by sl4m01** [reply]
Hello fvanroie,

thank you for your answer!
I checked my wiring and unfortunately i do have a slightly different wiring than described in the manual.
However, its also working using the TFT_eSPI library.

The following connections are different:
TFT / my wiring / OpenHASP manual
cs / g15 / g26
reset / g4 / g15
dc / g2 / g5
led / g16 / g21
t_cs / g21 / g17

Due to the fact that its already soldered and put in a 3D printed case I dont want to completely re-do it.

I tried to do my own compilation online with Gitpod.
However, i dont understand how to change the configured wiring (if even possible).

In the platformio_override-template.ini i found the following code:

debug_tool = esp-prog
debug_init_break = tbreak setup
build_flags =
    ${env.build_flags}
    ${esp32.build_flags}
;region -- TFT_eSPI build options -----------------------------------
    ${lcd.lolin24}
    ${esp32.vspi}        ; Use VSPI hardware SPI bus
    -D TFT_DC=5
    -D TFT_CS=26 
    -D TFT_RST=-1        ; RST
    -D TFT_BCKL=22       ; None, configurable via web UI (e.g. 21)
    -D TOUCH_CS=17       ; (can also be 22 or 16)
; -- Options ----------------------------------------
    -D HASP_USE_TELN

However, it seems to be for a different hardware setup (lcd.lolin24 and esp32.vspi).

Can you maybe give me a hint if its possible and where to adapt it?

EDIT:
I managed to get it done.
I opened the platformio_overwrite.ini and activated "d1-mini-esp32_ili9341".
Then i adapted the user_setups/esp32/d1-mini-esp32_ili9341.ini in the following way:

;region -- TFT_eSPI build options ------------------------ 
    ${lcd.lolin24}
    ${esp32.vspi}        ; Use VSPI hardware SPI bus
    -D TFT_DC=2          ; 5
    -D TFT_CS=15 
    -D TFT_RST=4        ; RST
    -D TFT_BCKL=16       ; None, configurable via web UI (e.g. 21)
    -D TOUCH_CS=21       ; (can also be 22 or 16)

Compiled and flashed it via Tasmota PyFlasher and its working :)

**Reply by fvanroie** [reply]
Good to hear you figured it out! There are actually several ways to do this, but adapting an existing config is certainly the easiest route.
